include/zephyr/senss/senss_sensor.h

```diff
@@ -0,0 +1,594 @@
+		const struct device *dev,
+		enum senss_sensor_state *state);
+
+/**
+ * @brief Trigger the data ready event to senss
+ *
+ * @param dev Pointer to the sensor device
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_sensor_notify_data_ready(
+		const struct device *dev);
+
+/**
+ * @brief Set the data ready mode of the sensor
+ *
+ * @param dev Pointer to the sensor device
+ *
+ * @param data_ready Enable/disable the data ready mode. Default:disabled
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_sensor_set_data_ready(
+		const struct device *dev, bool data_ready);
+
+/**
+ * @}
+ */
+
+/**
+ * @brief Sensor Callbacks
+ * @addtogroup senss_sensor_callbacks
+ * \{
+ */
+
+/**
+ * @brief Sensor initialize callback.
+ *
+ * Sensor can initialize it's runtime context in this callback.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param info The sensor instance's constant information.
+ *
+ * @param reporter_handles The reporters handles for this sensor, NULL for physical sensors.
+ *
+ * @param reporters_count The number of reporters, zero for physical sensors.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+typedef int (*senss_sensor_init_t)(
+		const struct device *dev, const struct senss_sensor_info *info,
+		const int *reporter_handles, int reporters_count);
+
+/**
+ * @brief Sensor's de-initialize callback.
+ *
+ * Sensor can release it's runtime context in this callback.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+typedef int (*senss_sensor_deinit_t)(
+		const struct device *dev);
+
+/**
+ * @brief Sensor reset callback.
+ *
+ * Sensor can reset its runtime context in this callback to default values without resources
+ * release and re-allocation.
+ *
+ * Its very useful for a virtual sensor to quickly reset its runtime context to a default state.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+typedef int (*senss_sensor_reset_t)(
+		const struct device *dev);
+
+/**
+ * @brief Sensor read sample callback.
+ *
+ * Only physical sensor need implement this callback.
+ * Physical sensor can fetch sample data from sensor device in this callback
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param buf Sensor subsystem runtime allocated buffer, and passed its pointer
+ * to this sensor for store fetched sample.
+ *
+ * @param size The size of the buffer in bytes.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+typedef int (*senss_sensor_read_sample_t)(
+		const struct device *dev,
+		void *buf, int size);
+
+/**
+ * @brief Sensor process data callback.
+ *
+ * Only virtual sensor need implement this callback.
+ * Virtual sensor can receive reporter's data and do fusion computing
+ * in this callback.
```

Comment:
Why can't virtual sensors just be clients of the subsystem? If they open a connection, they can set their callbacks, request sample rates, etc. The subsystem would arbitrate their requests like every other client. It would be much simpler to have the virtual sensors as both clients and producers.